From ddb610efd263421f3c6b4f4026d92dfbb1fb34c2 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Wed, 26 Mar 2025 14:19:39 -0500
Subject: [PATCH 01/10] Move LinalgExt::SortOp to LLVMGPUTileAndFuse pipeline
 (#20358)

The LinalgExt::SortOp is one of several ops that are handled by
LLVMGPUDistribute. This pipeline mismanages thread distribution resulting in racing
memory accesses which are causing certain tests to become flaky. So this
op is being moved to the more robust LLVMGPUTileAndFuse pipeline.

---------

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Dialect/GPU/TargetUtils/ConfigUtils.cpp   | 68 +++++++++++++++++++
 .../Dialect/GPU/TargetUtils/ConfigUtils.h     |  4 ++
 .../compiler/Codegen/LLVMGPU/KernelConfig.cpp | 49 +++++++++++++
 3 files changed, 121 insertions(+)

diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index b1e9ffa2dd25..877145f2ae7d 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -5,6 +5,7 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h"
+#include <cstdint>
 
 #include "iree/compiler/Codegen/Common/GPU/GPUHeuristics.h"
 #include "iree/compiler/Codegen/Common/TileInferenceUtils.h"
@@ -14,6 +15,7 @@
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUEnums.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.h"
 #include "iree/compiler/Codegen/Utils/Utils.h"
 #include "iree/compiler/Dialect/LinalgExt/Utils/Utils.h"
 #include "llvm/ADT/STLExtras.h"
@@ -996,6 +998,72 @@ LogicalResult setScatterLoweringConfig(IREE::GPU::TargetAttr target,
       {flatWorkgroupSize, 1, 1}, flatWorkgroupSize, DictionaryAttr());
 }
 
+//====---------------------------------------------------------------------===//
+// Sort Pipeline Configuration
+//====---------------------------------------------------------------------===//
+
+LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
+                            mlir::FunctionOpInterface entryPoint,
+                            Operation *op) {
+  MLIRContext *context = op->getContext();
+  Builder b(context);
+
+  auto subgroupSize = target.getPreferredSubgroupSize();
+  auto interfaceOp = cast<PartitionableLoopsInterface>(*op);
+  SmallVector<unsigned> partitionedLoops =
+      interfaceOp.getPartitionableLoops(kNumMaxParallelDims);
+
+  auto createLoweringConfig = [&](SmallVector<int64_t> &workgroupSizes,
+                                  SmallVector<int64_t> &threadSizes) {
+    SmallVector<NamedAttribute, 2> attrs = {
+        NamedAttribute("workgroup", b.getI64ArrayAttr(workgroupSizes)),
+        NamedAttribute("thread", b.getI64ArrayAttr(threadSizes))};
+    auto configDict = b.getDictionaryAttr(attrs);
+    return IREE::GPU::LoweringConfigAttr::get(context, configDict);
+  };
+
+  if (partitionedLoops.empty()) {
+    SmallVector<int64_t> defaultTileSizes(3, 1);
+    IREE::GPU::LoweringConfigAttr loweringConfig =
+        createLoweringConfig(defaultTileSizes, defaultTileSizes);
+    return setOpConfigAndEntryPointFnTranslation(
+        entryPoint, op, loweringConfig,
+        IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUTileAndFuse,
+        defaultTileSizes, subgroupSize, DictionaryAttr());
+  }
+
+  size_t numLoops = partitionedLoops.back() + 1;
+
+  // To get peak occupancy we need a workgroup size of at least two warps.
+  std::array<int64_t, 3> workgroupSize = {2 * subgroupSize, 1, 1};
+  SmallVector<int64_t> workgroupTileSizes(numLoops, 1);
+  SmallVector<int64_t> threadTileSizes(numLoops, 1);
+
+  // Set all non-parallel loops to zero tile size.
+  llvm::DenseSet<unsigned> partitionedLoopsSet(partitionedLoops.begin(),
+                                               partitionedLoops.end());
+  for (auto depth : llvm::seq<int64_t>(0, numLoops)) {
+    if (!partitionedLoopsSet.count(depth)) {
+      workgroupTileSizes[depth] = 0;
+    }
+  }
+
+  // Tile to have one element per thread.
+  for (int64_t depth = numLoops; depth > 0; --depth) {
+    if (partitionedLoopsSet.contains(depth - 1)) {
+      workgroupTileSizes[depth - 1] = workgroupSize[0];
+      break;
+    }
+  }
+
+  IREE::GPU::LoweringConfigAttr loweringConfig =
+      createLoweringConfig(workgroupTileSizes, threadTileSizes);
+  return setOpConfigAndEntryPointFnTranslation(
+      entryPoint, op, loweringConfig,
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUTileAndFuse,
+      workgroupSize, subgroupSize, DictionaryAttr());
+}
+
 //===----------------------------------------------------------------------===//
 // Lowering Config Attributes
 //===----------------------------------------------------------------------===//
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
index ad9aa426d6ed..9ccaa237cf69 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
@@ -47,6 +47,10 @@ LogicalResult setScatterLoweringConfig(IREE::GPU::TargetAttr target,
                                        mlir::FunctionOpInterface entryPoint,
                                        Operation *op);
 
+LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
+                            mlir::FunctionOpInterface entryPoint,
+                            Operation *op);
+
 //===----------------------------------------------------------------------===//
 // Pass Pipeline Options
 //===----------------------------------------------------------------------===//
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp b/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
index f9a91f7b69d8..3576a266d720 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
@@ -1956,6 +1956,55 @@ static LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
       workgroupSize);
 }
 
+//====---------------------------------------------------------------------===//
+// Pack/Unpack Pipeline Configuration
+//====---------------------------------------------------------------------===//
+
+static SmallVector<int64_t>
+getDefaultWorkgroupTileSizesForPackUnPack(TilingInterface op,
+                                          int64_t defaultSize) {
+  unsigned numLoops = op.getLoopIteratorTypes().size();
+  auto partitionedLoops = cast<PartitionableLoopsInterface>(op.getOperation())
+                              .getPartitionableLoops(kNumMaxParallelDims);
+  SmallVector<int64_t> workgroupTileSizes(numLoops, defaultSize);
+  llvm::DenseSet<unsigned> partitionedLoopsSet(partitionedLoops.begin(),
+                                               partitionedLoops.end());
+  for (auto dim : llvm::seq<int64_t>(0, workgroupTileSizes.size())) {
+    if (!partitionedLoopsSet.count(dim)) {
+      workgroupTileSizes[dim] = 0;
+    }
+  }
+
+  return workgroupTileSizes;
+}
+
+static LogicalResult setPackConfig(IREE::GPU::TargetAttr target,
+                                   mlir::FunctionOpInterface entryPoint,
+                                   linalg::PackOp packOp) {
+  SmallVector<int64_t> tileSizes = getDefaultWorkgroupTileSizesForPackUnPack(
+      cast<TilingInterface>(packOp.getOperation()),
+      target.getPreferredSubgroupSize());
+
+  // The default function aims to returns the number of workload per workgroup,
+  // but it does not know that it is working on packed domain. We need to take
+  // inner tile sizes into account and adjust the distribution tile sizes.
+  SmallVector<int64_t> innerTiles = packOp.getStaticTiles();
+  ArrayRef<int64_t> dimPos = packOp.getInnerDimsPos();
+  for (auto [pos, size] : llvm::zip_equal(dimPos, innerTiles)) {
+    if (tileSizes[pos] == 0 || ShapedType::isDynamic(size))
+      continue;
+    tileSizes[pos] = tileSizes[pos] / size;
+    tileSizes[pos] = std::max<int64_t>(tileSizes[pos], 1);
+  }
+
+  TileSizesListType tileSizesList = {tileSizes};
+  std::array<int64_t, 3> workgroupSizes = {target.getPreferredSubgroupSize(), 1,
+                                           1};
+  return setOpConfigAndEntryPointFnTranslation(
+      entryPoint, packOp, tileSizesList, CodeGenPipeline::LLVMGPUPackUnPack,
+      workgroupSizes);
+}
+
 //====---------------------------------------------------------------------===//
 // Default Pipeline Configuration
 //====---------------------------------------------------------------------===//

From 44c0b48a2d36a477cdb729c06c1767db57bb9936 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 13:02:18 -0500
Subject: [PATCH 02/10] changing tile sizes when no partitioned loops

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Dialect/GPU/TargetUtils/ConfigUtils.cpp   | 26 ++++++++++++++-----
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index 877145f2ae7d..7d437dc4e205 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -5,7 +5,6 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h"
-#include <cstdint>
 
 #include "iree/compiler/Codegen/Common/GPU/GPUHeuristics.h"
 #include "iree/compiler/Codegen/Common/TileInferenceUtils.h"
@@ -1005,6 +1004,8 @@ LogicalResult setScatterLoweringConfig(IREE::GPU::TargetAttr target,
 LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
                             mlir::FunctionOpInterface entryPoint,
                             Operation *op) {
+
+  assert(isa<IREE::LinalgExt::SortOp>(op) && "expected linalg_ext.sort op");
   MLIRContext *context = op->getContext();
   Builder b(context);
 
@@ -1023,13 +1024,13 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
   };
 
   if (partitionedLoops.empty()) {
-    SmallVector<int64_t> defaultTileSizes(3, 1);
+    SmallVector<int64_t> defaultTileSizes = {0};
     IREE::GPU::LoweringConfigAttr loweringConfig =
         createLoweringConfig(defaultTileSizes, defaultTileSizes);
     return setOpConfigAndEntryPointFnTranslation(
         entryPoint, op, loweringConfig,
         IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUTileAndFuse,
-        defaultTileSizes, subgroupSize, DictionaryAttr());
+        {1, 1, 1}, subgroupSize, DictionaryAttr());
   }
 
   size_t numLoops = partitionedLoops.back() + 1;
@@ -1045,13 +1046,24 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
   for (auto depth : llvm::seq<int64_t>(0, numLoops)) {
     if (!partitionedLoopsSet.count(depth)) {
       workgroupTileSizes[depth] = 0;
+      threadTileSizes[depth] = 0;
     }
   }
 
-  // Tile to have one element per thread.
-  for (int64_t depth = numLoops; depth > 0; --depth) {
-    if (partitionedLoopsSet.contains(depth - 1)) {
-      workgroupTileSizes[depth - 1] = workgroupSize[0];
+  // Tile to have one element per thread
+  ArrayRef loopBounds = cast<IREE::LinalgExt::SortOp>(op).getOperandShape();
+  int64_t residualWorkgroupSize = workgroupSize[0];
+  for (int64_t depth = numLoops - 1; depth >= 0; --depth) {
+    if (!partitionedLoopsSet.contains(depth)) {
+      continue;
+    }
+    if (residualWorkgroupSize % loopBounds[depth] == 0) {
+      workgroupTileSizes[depth] = loopBounds[depth];
+      residualWorkgroupSize /= loopBounds[depth];
+      continue;
+    }
+    if (loopBounds[depth] % residualWorkgroupSize == 0) {
+      workgroupTileSizes[depth] = residualWorkgroupSize;
       break;
     }
   }

From cfefcb00d308cced53409268b81ad37a138165b5 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 15:08:30 -0500
Subject: [PATCH 03/10] updating distribution verificaiton to skip for
 workgroup size 1

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Common/GPU/GPUVerifyDistribution.cpp        | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
index a04876f00057..d261ff22bd66 100644
--- a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
+++ b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
@@ -29,6 +29,23 @@ struct GPUVerifyDistributionPass final
 
   void runOnOperation() override {
     FunctionOpInterface funcOp = getOperation();
+    
+    std::optional<SmallVector<int64_t>> workgroupSize =
+        getWorkgroupSize(funcOp);
+    if (!workgroupSize) {
+      funcOp->emitOpError("requires a workgroup size attribute.");
+      return signalPassFailure();
+    }
+
+    bool multithreaded = false;
+    for (int64_t size : *workgroupSize) {
+      if (size != 1) {
+        multithreaded = true;
+        break;
+      }
+    }
+    if (!multithreaded)
+      return;
 
     auto privateAddressSpace = gpu::AddressSpaceAttr::get(
         &getContext(), gpu::GPUDialect::getPrivateAddressSpace());

From 148d3224114f3c404ec5b10d52d1b02134039442 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 15:08:57 -0500
Subject: [PATCH 04/10] adding sortConfig test

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Codegen/LLVMGPU/test/config_sort.mlir     | 91 +++++++++++++++++++
 1 file changed, 91 insertions(+)
 create mode 100644 compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir

diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
new file mode 100644
index 000000000000..07263d41f5be
--- /dev/null
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
@@ -0,0 +1,91 @@
+// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 --pass-pipeline='builtin.module(iree-llvmgpu-select-lowering-strategy)' %s | FileCheck %s
+
+module {
+  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
+    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    return
+  }
+}
+
+//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [1, 1, 1] subgroup_size = 32>
+//       CHECK: func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store()
+//       CHECK:     translation_info = #[[$TRANSLATION]]
+//       CHECK:   iree_linalg_ext.sort {
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
+
+// -----
+
+module {
+  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>> -> tensor<2x4xi32>
+    %2 = iree_linalg_ext.sort dimension(1) outs(%1 : tensor<2x4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<2x4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : tensor<2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
+    return
+  }
+}
+
+//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+//       CHECK: func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store()
+//       CHECK:     translation_info = #[[$TRANSLATION]]
+//       CHECK:   iree_linalg_ext.sort {
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1], workgroup = [2]}>
+
+// -----
+
+module {
+  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>> -> tensor<1x2x4xi32>
+    %2 = iree_linalg_ext.sort dimension(2) outs(%1 : tensor<1x2x4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<1x2x4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : tensor<1x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
+    return
+  }
+}
+
+//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+//       CHECK: func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store()
+//       CHECK:     translation_info = #[[$TRANSLATION]]
+//       CHECK:   iree_linalg_ext.sort {
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1, 1], workgroup = [1, 2]}>
+
+// -----
+
+module {
+  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
+    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi sgt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    return
+  }
+}
+
+//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [1, 1, 1] subgroup_size = 32>
+//       CHECK-DAG: func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store()
+//       CHECK:     translation_info = #[[$TRANSLATION]]
+//       CHECK:   iree_linalg_ext.sort {
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
\ No newline at end of file

From f94a374dfacb10ec6ed924692557fa5ec82cf6f5 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 15:15:12 -0500
Subject: [PATCH 05/10] meeting coding standards

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Codegen/Common/GPU/GPUVerifyDistribution.cpp      |  5 +++--
 .../Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp   | 11 +++++------
 .../compiler/Codegen/LLVMGPU/test/config_sort.mlir    |  2 +-
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
index d261ff22bd66..a7adc2f90dd7 100644
--- a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
+++ b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
@@ -29,7 +29,7 @@ struct GPUVerifyDistributionPass final
 
   void runOnOperation() override {
     FunctionOpInterface funcOp = getOperation();
-    
+
     std::optional<SmallVector<int64_t>> workgroupSize =
         getWorkgroupSize(funcOp);
     if (!workgroupSize) {
@@ -44,8 +44,9 @@ struct GPUVerifyDistributionPass final
         break;
       }
     }
-    if (!multithreaded)
+    if (!multithreaded) {
       return;
+    }
 
     auto privateAddressSpace = gpu::AddressSpaceAttr::get(
         &getContext(), gpu::GPUDialect::getPrivateAddressSpace());
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index 7d437dc4e205..593595ee2ad2 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -1014,9 +1014,9 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
   SmallVector<unsigned> partitionedLoops =
       interfaceOp.getPartitionableLoops(kNumMaxParallelDims);
 
-  auto createLoweringConfig = [&](SmallVector<int64_t> &workgroupSizes,
-                                  SmallVector<int64_t> &threadSizes) {
-    SmallVector<NamedAttribute, 2> attrs = {
+  auto createLoweringConfig = [&](ArrayRef<int64_t> workgroupSizes,
+                                  ArrayRef<int64_t> threadSizes) {
+    NamedAttribute attrs[2] = {
         NamedAttribute("workgroup", b.getI64ArrayAttr(workgroupSizes)),
         NamedAttribute("thread", b.getI64ArrayAttr(threadSizes))};
     auto configDict = b.getDictionaryAttr(attrs);
@@ -1024,9 +1024,8 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
   };
 
   if (partitionedLoops.empty()) {
-    SmallVector<int64_t> defaultTileSizes = {0};
     IREE::GPU::LoweringConfigAttr loweringConfig =
-        createLoweringConfig(defaultTileSizes, defaultTileSizes);
+        createLoweringConfig(int64_t{0}, int64_t{0});
     return setOpConfigAndEntryPointFnTranslation(
         entryPoint, op, loweringConfig,
         IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUTileAndFuse,
@@ -1050,7 +1049,7 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
     }
   }
 
-  // Tile to have one element per thread
+  // Tile to have one element per thread.
   ArrayRef loopBounds = cast<IREE::LinalgExt::SortOp>(op).getOperandShape();
   int64_t residualWorkgroupSize = workgroupSize[0];
   for (int64_t depth = numLoops - 1; depth >= 0; --depth) {
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
index 07263d41f5be..988039443d6c 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
@@ -88,4 +88,4 @@ module {
 //       CHECK-DAG: func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
-//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
\ No newline at end of file
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>

From 7ec3cb7cf45e5f7da298961c42855f53f5dc9d71 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 17:16:47 -0500
Subject: [PATCH 06/10] adding pipeline test

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../compiler/Codegen/LLVMGPU/test/BUILD.bazel |  2 +
 .../Codegen/LLVMGPU/test/CMakeLists.txt       |  2 +
 .../LLVMGPU/test/sort_pipeline_test.mlir      | 87 +++++++++++++++++++
 3 files changed, 91 insertions(+)
 create mode 100644 compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir

diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/BUILD.bazel b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/BUILD.bazel
index a3a448222c79..c4cf0c07a8fd 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/BUILD.bazel
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/BUILD.bazel
@@ -32,6 +32,7 @@ iree_lit_test_suite(
             "config_horizontally_fused_ops.mlir",
             "config_matvec.mlir",
             "config_root_op_attribute.mlir",
+            "config_sort.mlir",
             "config_winograd.mlir",
             "extract_address_computation_gpu.mlir",
             "gpu_set_num_workgroups.mlir",
@@ -54,6 +55,7 @@ iree_lit_test_suite(
             "pack_shared_memory_alloc.mlir",
             "prefetch_shared_memory.mlir",
             "promote_matmul_to_fit_mma.mlir",
+            "sort_pipeline_test.mlir",
             "tensor_pad.mlir",
             "tensorcore_vectorization.mlir",
             "test_query_mma.mlir",
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/CMakeLists.txt b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/CMakeLists.txt
index a1012ed8c9ab..fce874248c95 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/CMakeLists.txt
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/CMakeLists.txt
@@ -21,6 +21,7 @@ iree_lit_test_suite(
     "config_horizontally_fused_ops.mlir"
     "config_matvec.mlir"
     "config_root_op_attribute.mlir"
+    "config_sort.mlir"
     "config_winograd.mlir"
     "configure_tensor_layout.mlir"
     "conv_pipeline_test_cuda.mlir"
@@ -51,6 +52,7 @@ iree_lit_test_suite(
     "reduction_pipeline_softmax_rocm.mlir"
     "reuse_shared_memory_allocs.mlir"
     "rocdl_pipeline_test.mlir"
+    "sort_pipeline_test.mlir"
     "tensor_pad.mlir"
     "tensorcore_vectorization.mlir"
     "test_query_mma.mlir"
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
new file mode 100644
index 000000000000..b2d9b70e37bf
--- /dev/null
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
@@ -0,0 +1,87 @@
+// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 --pass-pipeline="builtin.module(iree-llvmgpu-select-lowering-strategy, func.func(iree-llvmgpu-lower-executable-target))" /home/muzasyed/iree/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir | FileCheck /home/muzasyed/iree/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
+
+module {
+  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
+    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    return
+  }
+}
+
+//   CHECK-LABEL:  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store
+//         CHECK:      amdgpu.fat_raw_buffer_cast
+//         CHECK:      memref.assume_alignment
+
+// -----
+
+module {
+  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>> -> tensor<2x4xi32>
+    %2 = iree_linalg_ext.sort dimension(1) outs(%1 : tensor<2x4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<2x4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : tensor<2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
+    return
+  }
+}
+
+//   CHECK-LABEL:  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store
+//         CHECK:      amdgpu.fat_raw_buffer_cast
+//         CHECK:      memref.assume_alignment
+//         CHECK:      scf.for
+//         CHECK:        memref.subview
+
+// -----
+
+module {
+  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>> -> tensor<1x2x4xi32>
+    %2 = iree_linalg_ext.sort dimension(2) outs(%1 : tensor<1x2x4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<1x2x4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : tensor<1x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
+    return
+  }
+}
+
+//   CHECK-LABEL:  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store
+//         CHECK:      amdgpu.fat_raw_buffer_cast
+//         CHECK:      memref.assume_alignment
+//         CHECK:      scf.for
+//         CHECK:        memref.subview
+
+// -----
+
+module {
+  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store() {
+    %c0 = arith.constant 0 : index
+    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
+    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+    ^bb0(%arg0: i32, %arg1: i32):
+      %3 = arith.cmpi sgt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %3 : i1
+    } -> tensor<4xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    return
+  }
+}
+
+//   CHECK-LABEL:  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store
+//         CHECK:      amdgpu.fat_raw_buffer_cast
+//         CHECK:      memref.assume_alignment

From a062d1732a19176eda4844ed85be28967ac0d154 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Fri, 28 Mar 2025 18:11:22 -0500
Subject: [PATCH 07/10] Addressing PR comments

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Common/GPU/GPUVerifyDistribution.cpp      |   9 +-
 .../GPU/test/gpu_verify_distribution.mlir     |  15 ++-
 .../Dialect/GPU/TargetUtils/ConfigUtils.cpp   |  25 ++--
 .../compiler/Codegen/LLVMGPU/KernelConfig.cpp |  51 +------
 .../Codegen/LLVMGPU/test/config_sort.mlir     | 124 +++++++++---------
 .../LLVMGPU/test/gpu_set_num_workgroups.mlir  |   5 +-
 .../LLVMGPU/test/sort_pipeline_test.mlir      |  87 ++++++------
 7 files changed, 128 insertions(+), 188 deletions(-)

diff --git a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
index a7adc2f90dd7..fa20061e9d02 100644
--- a/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
+++ b/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVerifyDistribution.cpp
@@ -37,14 +37,7 @@ struct GPUVerifyDistributionPass final
       return signalPassFailure();
     }
 
-    bool multithreaded = false;
-    for (int64_t size : *workgroupSize) {
-      if (size != 1) {
-        multithreaded = true;
-        break;
-      }
-    }
-    if (!multithreaded) {
+    if (llvm::all_of(*workgroupSize, [](int64_t size) { return size == 1; })) {
       return;
     }
 
diff --git a/compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_verify_distribution.mlir b/compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_verify_distribution.mlir
index c7f45698fb95..b57f3e668c15 100644
--- a/compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_verify_distribution.mlir
+++ b/compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_verify_distribution.mlir
@@ -1,7 +1,17 @@
 // RUN: iree-opt %s --split-input-file --verify-diagnostics \
 // RUN:   --pass-pipeline="builtin.module(func.func(iree-codegen-gpu-verify-distribution))"
 
-func.func @unmapped_forall(%out : memref<32xi32>) {
+// expected-error @+1 {{requires a workgroup size attribute}}
+func.func @incomplete_funcop(%out : memref<32xi32>) {
+  scf.forall (%arg0) in (32) {
+  }
+  return
+}
+
+// -----
+
+#translation = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+func.func @unmapped_forall(%out : memref<32xi32>) attributes {translation_info = #translation} {
   // expected-error @+1 {{requires a mapping attribute}}
   scf.forall (%arg0) in (32) {
   }
@@ -10,7 +20,8 @@ func.func @unmapped_forall(%out : memref<32xi32>) {
 
 // -----
 
-func.func @lane_forall_no_warp_parent(%out : memref<32xi32>) {
+#translation = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+func.func @lane_forall_no_warp_parent(%out : memref<32xi32>) attributes {translation_info = #translation} {
   // expected-error@+1 {{lane distributed scf.forall must have a parent subgroup distributed loop}}
   scf.forall (%arg0) in (32) {
   } {mapping = [#iree_gpu.lane_id<0>]}
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index 593595ee2ad2..9c5a310891fb 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -5,6 +5,7 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h"
+#include <optional>
 
 #include "iree/compiler/Codegen/Common/GPU/GPUHeuristics.h"
 #include "iree/compiler/Codegen/Common/TileInferenceUtils.h"
@@ -621,18 +622,6 @@ LogicalResult setTileAndFuseLoweringConfig(IREE::GPU::TargetAttr target,
   if (failed(maybeDistInfo)) {
     return failure();
   }
-
-  // TODO(Max191): Drop this check for reshapes in the dispatch once we can
-  // codegen larger tile sizes with reshapes in the dispatch.
-  bool hasReshapes = false;
-  entryPoint->walk([&](Operation *opInEntryPoint) {
-    if (isa<tensor::ExpandShapeOp, tensor::CollapseShapeOp>(opInEntryPoint)) {
-      hasReshapes = true;
-      return WalkResult::interrupt();
-    }
-    return WalkResult::advance();
-  });
-
   DistributionInfo distInfo = maybeDistInfo.value();
 
   const int subgroupSize = target.getPreferredSubgroupSize();
@@ -752,7 +741,7 @@ LogicalResult setTileAndFuseLoweringConfig(IREE::GPU::TargetAttr target,
         // TODO: Try to take into account element type bit width to get
         // 4xdword reads instead of 4x{elements}.
         if (distInfo.vectorizable && wgDim == 0 && !lossFactor &&
-            candidate % numVectorElements == 0 && !hasReshapes) {
+            candidate % numVectorElements == 0) {
           // Use size-1 vectors to increase parallelism if larger ones causes
           // idle threads in the subgroup.
           bool hasIdleThreads = distInfo.partitionableLoops.size() == 1 &&
@@ -1004,15 +993,14 @@ LogicalResult setScatterLoweringConfig(IREE::GPU::TargetAttr target,
 LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
                             mlir::FunctionOpInterface entryPoint,
                             Operation *op) {
-
   assert(isa<IREE::LinalgExt::SortOp>(op) && "expected linalg_ext.sort op");
   MLIRContext *context = op->getContext();
   Builder b(context);
 
-  auto subgroupSize = target.getPreferredSubgroupSize();
+  const int64_t subgroupSize = target.getPreferredSubgroupSize();
   auto interfaceOp = cast<PartitionableLoopsInterface>(*op);
   SmallVector<unsigned> partitionedLoops =
-      interfaceOp.getPartitionableLoops(kNumMaxParallelDims);
+      interfaceOp.getPartitionableLoops(std::nullopt);
 
   auto createLoweringConfig = [&](ArrayRef<int64_t> workgroupSizes,
                                   ArrayRef<int64_t> threadSizes) {
@@ -1032,7 +1020,7 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
         {1, 1, 1}, subgroupSize, DictionaryAttr());
   }
 
-  size_t numLoops = partitionedLoops.back() + 1;
+  unsigned numLoops = cast<ShapedType>(op->getResult(0).getType()).getRank();
 
   // To get peak occupancy we need a workgroup size of at least two warps.
   std::array<int64_t, 3> workgroupSize = {2 * subgroupSize, 1, 1};
@@ -1056,6 +1044,9 @@ LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
     if (!partitionedLoopsSet.contains(depth)) {
       continue;
     }
+    if (ShapedType::isDynamic(loopBounds[depth])) {
+      continue;
+    }
     if (residualWorkgroupSize % loopBounds[depth] == 0) {
       workgroupTileSizes[depth] = loopBounds[depth];
       residualWorkgroupSize /= loopBounds[depth];
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp b/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
index 3576a266d720..607eb6fef112 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp
@@ -1956,55 +1956,6 @@ static LogicalResult setSortConfig(IREE::GPU::TargetAttr target,
       workgroupSize);
 }
 
-//====---------------------------------------------------------------------===//
-// Pack/Unpack Pipeline Configuration
-//====---------------------------------------------------------------------===//
-
-static SmallVector<int64_t>
-getDefaultWorkgroupTileSizesForPackUnPack(TilingInterface op,
-                                          int64_t defaultSize) {
-  unsigned numLoops = op.getLoopIteratorTypes().size();
-  auto partitionedLoops = cast<PartitionableLoopsInterface>(op.getOperation())
-                              .getPartitionableLoops(kNumMaxParallelDims);
-  SmallVector<int64_t> workgroupTileSizes(numLoops, defaultSize);
-  llvm::DenseSet<unsigned> partitionedLoopsSet(partitionedLoops.begin(),
-                                               partitionedLoops.end());
-  for (auto dim : llvm::seq<int64_t>(0, workgroupTileSizes.size())) {
-    if (!partitionedLoopsSet.count(dim)) {
-      workgroupTileSizes[dim] = 0;
-    }
-  }
-
-  return workgroupTileSizes;
-}
-
-static LogicalResult setPackConfig(IREE::GPU::TargetAttr target,
-                                   mlir::FunctionOpInterface entryPoint,
-                                   linalg::PackOp packOp) {
-  SmallVector<int64_t> tileSizes = getDefaultWorkgroupTileSizesForPackUnPack(
-      cast<TilingInterface>(packOp.getOperation()),
-      target.getPreferredSubgroupSize());
-
-  // The default function aims to returns the number of workload per workgroup,
-  // but it does not know that it is working on packed domain. We need to take
-  // inner tile sizes into account and adjust the distribution tile sizes.
-  SmallVector<int64_t> innerTiles = packOp.getStaticTiles();
-  ArrayRef<int64_t> dimPos = packOp.getInnerDimsPos();
-  for (auto [pos, size] : llvm::zip_equal(dimPos, innerTiles)) {
-    if (tileSizes[pos] == 0 || ShapedType::isDynamic(size))
-      continue;
-    tileSizes[pos] = tileSizes[pos] / size;
-    tileSizes[pos] = std::max<int64_t>(tileSizes[pos], 1);
-  }
-
-  TileSizesListType tileSizesList = {tileSizes};
-  std::array<int64_t, 3> workgroupSizes = {target.getPreferredSubgroupSize(), 1,
-                                           1};
-  return setOpConfigAndEntryPointFnTranslation(
-      entryPoint, packOp, tileSizesList, CodeGenPipeline::LLVMGPUPackUnPack,
-      workgroupSizes);
-}
-
 //====---------------------------------------------------------------------===//
 // Default Pipeline Configuration
 //====---------------------------------------------------------------------===//
@@ -2778,7 +2729,7 @@ static LogicalResult setRootConfig(IREE::GPU::TargetAttr target,
       })
       .Case<IREE::LinalgExt::SortOp>([&](auto sortOp) {
         LDBG("Sort Config");
-        return setSortConfig(target, entryPointFn, sortOp);
+        return IREE::GPU::setSortConfig(target, entryPointFn, sortOp);
       })
       .Case<IREE::LinalgExt::WinogradInputTransformOp,
             IREE::LinalgExt::WinogradOutputTransformOp,
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
index 988039443d6c..1d2cee51b8b2 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
@@ -1,91 +1,89 @@
-// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 --pass-pipeline='builtin.module(iree-llvmgpu-select-lowering-strategy)' %s | FileCheck %s
+// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 --pass-pipeline='builtin.module(iree-llvmgpu-select-lowering-strategy)' %s | \
+// RUN: FileCheck %s
 
-module {
-  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store() {
-    %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
-    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
-    ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi slt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
-    return
-  }
+func.func @sort1D() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+  %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
+  %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+  ^bb0(%arg0: i32, %arg1: i32):
+    %3 = arith.cmpi slt, %arg0, %arg1 : i32
+    iree_linalg_ext.yield %3 : i1
+  } -> tensor<4xi32>
+  flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+  return
 }
 
+
 //       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [1, 1, 1] subgroup_size = 32>
-//       CHECK: func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store()
+//       CHECK: func.func @sort1D()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
 //  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
 
 // -----
 
-module {
-  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store() {
-    %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>> -> tensor<2x4xi32>
-    %2 = iree_linalg_ext.sort dimension(1) outs(%1 : tensor<2x4xi32>) {
-    ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi slt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<2x4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : tensor<2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
-    return
-  }
+func.func @sort2D_static_shape() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>>
+  %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2000, 30000], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>> -> tensor<2000x30000xi32>
+  %2 = iree_linalg_ext.sort dimension(1) outs(%1 : tensor<2000x30000xi32>) {
+  ^bb0(%arg0: i32, %arg1: i32):
+    %3 = arith.cmpi slt, %arg0, %arg1 : i32
+    iree_linalg_ext.yield %3 : i1
+  } -> tensor<2000x30000xi32>
+  flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2000, 30000], strides = [1, 1] : tensor<2000x30000xi32> -> !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>>
+  return
 }
 
 //       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
-//       CHECK: func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store()
+//       CHECK: func.func @sort2D_static_shape()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
-//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1], workgroup = [2]}>
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1, 0], workgroup = [1, 0]}>
 
 // -----
-
-module {
-  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store() {
-    %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>> -> tensor<1x2x4xi32>
-    %2 = iree_linalg_ext.sort dimension(2) outs(%1 : tensor<1x2x4xi32>) {
-    ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi slt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<1x2x4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : tensor<1x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
-    return
-  }
+#pipeline_layout = #hal.pipeline.layout<constants = 2, bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>
+func.func @sort3D_dynamic_shape() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.constant.load layout(#pipeline_layout) ordinal(0) : i32
+  %1 = hal.interface.constant.load layout(#pipeline_layout) ordinal(1) : i32
+  %2 = arith.index_castui %0 : i32 to index
+  %3 = arith.index_castui %1 : i32 to index
+  %4 = flow.dispatch.workload.ordinal %3, 0 : index
+  %5 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%2) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4}
+  %6 = flow.dispatch.tensor.load %5, offsets = [0, 0, 0], sizes = [%4, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4} -> tensor<?x2x4xi32>
+  %7 = iree_linalg_ext.sort dimension(2) outs(%6 : tensor<?x2x4xi32>) {
+  ^bb0(%arg0: i32, %arg1: i32):
+    %8 = arith.cmpi slt, %arg0, %arg1 : i32
+    iree_linalg_ext.yield %8 : i1
+  } -> tensor<?x2x4xi32>
+  flow.dispatch.tensor.store %7, %5, offsets = [0, 0, 0], sizes = [%4, 2, 4], strides = [1, 1, 1] : tensor<?x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4}
+  return
 }
 
 //       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
-//       CHECK: func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store()
+//       CHECK: func.func @sort3D_dynamic_shape()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
-//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1, 1], workgroup = [1, 2]}>
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [1, 1, 0], workgroup = [1, 2, 0]}>
 
 // -----
-
-module {
-  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store() {
-    %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
-    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
-    ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi sgt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
-    return
-  }
+func.func @sort5D_static_shape() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4x100x100x200x300xi32>>
+  %1 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0, 0, 0], sizes = [4, 100, 100, 200, 300], strides = [1, 1, 1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<4x100x100x200x300xi32>> -> tensor<4x100x100x200x300xi32>
+  %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4x100x100x200x300xi32>) {
+  ^bb0(%arg0: i32, %arg1: i32):
+    %3 = arith.cmpi sgt, %arg0, %arg1 : i32
+    iree_linalg_ext.yield %3 : i1
+  } -> tensor<4x100x100x200x300xi32>
+  flow.dispatch.tensor.store %2, %0, offsets = [0, 0, 0, 0, 0], sizes = [4, 100, 100, 200, 300], strides = [1, 1, 1, 1, 1] : tensor<4x100x100x200x300xi32> -> !flow.dispatch.tensor<readwrite:tensor<4x100x100x200x300xi32>>
+  return
 }
 
-//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [1, 1, 1] subgroup_size = 32>
-//       CHECK-DAG: func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store()
+//       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+//       CHECK-DAG: func.func @sort5D_static_shape()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
-//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
+//  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0, 1, 1, 1, 1], workgroup = [0, 1, 1, 1, 1]}>
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir
index 7950e0bfcb67..a14835508927 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir
@@ -304,12 +304,11 @@ func.func @sort_op() {
   return
 }
 
-//   CHECK-DAG: #[[CONFIG:.+]] = #iree_codegen.lowering_config<tile_sizes = {{\[}}[64]{{\]}}>
-//   CHECK-DAG: #[[TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUDistribute workgroup_size = [64, 1, 1]>
+//       CHECK: #[[TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
 //       CHECK: func.func @sort_op()
 //  CHECK-SAME:     translation_info = #[[TRANSLATION]]
 //       CHECK: iree_linalg_ext.sort
-//  CHECK-SAME:     lowering_config = #[[CONFIG]]
+//  CHECK-SAME:     lowering_config = #iree_gpu.lowering_config<{thread = [1, 0], workgroup = [1, 0]}>
 
 // -----
 
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
index b2d9b70e37bf..cbaa301bc6bd 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
@@ -1,11 +1,16 @@
-// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 --pass-pipeline="builtin.module(iree-llvmgpu-select-lowering-strategy, func.func(iree-llvmgpu-lower-executable-target))" /home/muzasyed/iree/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir | FileCheck /home/muzasyed/iree/compiler/src/iree/compiler/Codegen/LLVMGPU/test/sort_pipeline_test.mlir
+// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx1100 \
+// RUN:     --pass-pipeline="builtin.module(func.func(iree-llvmgpu-lower-executable-target))" %s | \
+// RUN: FileCheck %s
 
+#pipeline_layout = #hal.pipeline.layout<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>
+#translation = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [1, 1, 1] subgroup_size = 32>
+#lowering_config = #iree_gpu.lowering_config<{thread = [0], workgroup = [0]}>
 module {
-  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store() {
+  func.func @sort1D() attributes {translation_info = #translation} {
     %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+    %0 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
     %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
-    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
+    %2 = iree_linalg_ext.sort {lowering_config = #lowering_config} dimension(0) outs(%1 : tensor<4xi32>) {
     ^bb0(%arg0: i32, %arg1: i32):
       %3 = arith.cmpi slt, %arg0, %arg1 : i32
       iree_linalg_ext.yield %3 : i1
@@ -15,73 +20,65 @@ module {
   }
 }
 
-//   CHECK-LABEL:  func.func @_sort1D_dispatch_0_sort_4xi32_dispatch_tensor_store
+//   CHECK-LABEL:  func.func @sort1D
 //         CHECK:      amdgpu.fat_raw_buffer_cast
 //         CHECK:      memref.assume_alignment
 
 // -----
 
+#pipeline_layout = #hal.pipeline.layout<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>
+#translation = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+#lowering_config = #iree_gpu.lowering_config<{thread = [1], workgroup = [1]}>
 module {
-  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store() {
+  func.func @sort2D_static_shape() attributes {translation_info = #translation} {
     %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2x4xi32>> -> tensor<2x4xi32>
-    %2 = iree_linalg_ext.sort dimension(1) outs(%1 : tensor<2x4xi32>) {
+    %0 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>>
+    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>> -> tensor<2000x30000xi32>
+    %2 = iree_linalg_ext.sort {lowering_config = #lowering_config} dimension(1) outs(%1 : tensor<2000x30000xi32>) {
     ^bb0(%arg0: i32, %arg1: i32):
       %3 = arith.cmpi slt, %arg0, %arg1 : i32
       iree_linalg_ext.yield %3 : i1
-    } -> tensor<2x4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2, 4], strides = [1, 1] : tensor<2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<2x4xi32>>
+    } -> tensor<2000x30000xi32>
+    flow.dispatch.tensor.store %2, %0, offsets = [0, 0], sizes = [2000, 30000], strides = [1, 1] : tensor<2000x30000xi32> -> !flow.dispatch.tensor<readwrite:tensor<2000x30000xi32>>
     return
   }
 }
 
-//   CHECK-LABEL:  func.func @_sort2D_dispatch_0_sort_2x4xi32_dispatch_tensor_store
+//   CHECK-LABEL:  func.func @sort2D_static_shape
 //         CHECK:      amdgpu.fat_raw_buffer_cast
 //         CHECK:      memref.assume_alignment
-//         CHECK:      scf.for
+//         CHECK:      scf.forall
 //         CHECK:        memref.subview
+//         CHECK:        scf.for
 
 // -----
 
+#pipeline_layout = #hal.pipeline.layout<constants = 2, bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>
+#translation = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
+#lowering_config = #iree_gpu.lowering_config<{thread = [1, 1], workgroup = [1, 2]}>
 module {
-  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store() {
+  func.func @sort3D_dynamic_shape() attributes {translation_info = #translation} {
     %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>> -> tensor<1x2x4xi32>
-    %2 = iree_linalg_ext.sort dimension(2) outs(%1 : tensor<1x2x4xi32>) {
+    %0 = hal.interface.constant.load layout(#pipeline_layout) ordinal(0) : i32
+    %1 = hal.interface.constant.load layout(#pipeline_layout) ordinal(1) : i32
+    %2 = arith.index_castui %0 : i32 to index
+    %3 = arith.index_castui %1 : i32 to index
+    %4 = flow.dispatch.workload.ordinal %3, 0 : index
+    %5 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%2) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4}
+    %6 = flow.dispatch.tensor.load %5, offsets = [0, 0, 0], sizes = [%4, 2, 4], strides = [1, 1, 1] : !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4} -> tensor<?x2x4xi32>
+    %7 = iree_linalg_ext.sort {lowering_config = #lowering_config} dimension(2) outs(%6 : tensor<?x2x4xi32>) {
     ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi slt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<1x2x4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0, 0, 0], sizes = [1, 2, 4], strides = [1, 1, 1] : tensor<1x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<1x2x4xi32>>
-    return
-  }
-}
-
-//   CHECK-LABEL:  func.func @_sort3D_dispatch_0_sort_1x2x4xi32_dispatch_tensor_store
-//         CHECK:      amdgpu.fat_raw_buffer_cast
-//         CHECK:      memref.assume_alignment
-//         CHECK:      scf.for
-//         CHECK:        memref.subview
-
-// -----
-
-module {
-  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store() {
-    %c0 = arith.constant 0 : index
-    %0 = hal.interface.binding.subspan layout(<bindings = [#hal.pipeline.binding<storage_buffer, Indirect>], flags = Indirect>) binding(0) alignment(64) offset(%c0) flags(Indirect) {iree_gpu.use_rocdl_buffer_instructions} : !flow.dispatch.tensor<readwrite:tensor<4xi32>>
-    %1 = flow.dispatch.tensor.load %0, offsets = [0], sizes = [4], strides = [1] : !flow.dispatch.tensor<readwrite:tensor<4xi32>> -> tensor<4xi32>
-    %2 = iree_linalg_ext.sort dimension(0) outs(%1 : tensor<4xi32>) {
-    ^bb0(%arg0: i32, %arg1: i32):
-      %3 = arith.cmpi sgt, %arg0, %arg1 : i32
-      iree_linalg_ext.yield %3 : i1
-    } -> tensor<4xi32>
-    flow.dispatch.tensor.store %2, %0, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<readwrite:tensor<4xi32>>
+      %8 = arith.cmpi slt, %arg0, %arg1 : i32
+      iree_linalg_ext.yield %8 : i1
+    } -> tensor<?x2x4xi32>
+    flow.dispatch.tensor.store %7, %5, offsets = [0, 0, 0], sizes = [%4, 2, 4], strides = [1, 1, 1] : tensor<?x2x4xi32> -> !flow.dispatch.tensor<readwrite:tensor<?x2x4xi32>>{%4}
     return
   }
 }
 
-//   CHECK-LABEL:  func.func @_sort_to_decreasing_seq_dispatch_0_sort_4xi32_dispatch_tensor_store
+//   CHECK-LABEL:  func.func @sort3D_dynamic_shape
 //         CHECK:      amdgpu.fat_raw_buffer_cast
 //         CHECK:      memref.assume_alignment
+//         CHECK:      scf.forall
+//         CHECK:       scf.for
+//         CHECK:         memref.subview

From f4f17e46de64c5ada7ae293b8fdfa81cb784345e Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Mon, 7 Apr 2025 14:48:15 -0500
Subject: [PATCH 08/10] removing rebase error

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../Dialect/GPU/TargetUtils/ConfigUtils.cpp      | 16 +++++++++++++---
 .../Dialect/GPU/TargetUtils/ConfigUtils.h        |  1 +
 .../Codegen/LLVMGPU/test/config_sort.mlir        |  2 +-
 3 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index 9c5a310891fb..5c24338866fc 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -5,7 +5,6 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include "iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h"
-#include <optional>
 
 #include "iree/compiler/Codegen/Common/GPU/GPUHeuristics.h"
 #include "iree/compiler/Codegen/Common/TileInferenceUtils.h"
@@ -15,7 +14,6 @@
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUEnums.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
-#include "iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.h"
 #include "iree/compiler/Codegen/Utils/Utils.h"
 #include "iree/compiler/Dialect/LinalgExt/Utils/Utils.h"
 #include "llvm/ADT/STLExtras.h"
@@ -622,6 +620,18 @@ LogicalResult setTileAndFuseLoweringConfig(IREE::GPU::TargetAttr target,
   if (failed(maybeDistInfo)) {
     return failure();
   }
+
+  // TODO(Max191): Drop this check for reshapes in the dispatch once we can
+  // codegen larger tile sizes with reshapes in the dispatch.
+  bool hasReshapes = false;
+  entryPoint->walk([&](Operation *opInEntryPoint) {
+    if (isa<tensor::ExpandShapeOp, tensor::CollapseShapeOp>(opInEntryPoint)) {
+      hasReshapes = true;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+
   DistributionInfo distInfo = maybeDistInfo.value();
 
   const int subgroupSize = target.getPreferredSubgroupSize();
@@ -741,7 +751,7 @@ LogicalResult setTileAndFuseLoweringConfig(IREE::GPU::TargetAttr target,
         // TODO: Try to take into account element type bit width to get
         // 4xdword reads instead of 4x{elements}.
         if (distInfo.vectorizable && wgDim == 0 && !lossFactor &&
-            candidate % numVectorElements == 0) {
+            candidate % numVectorElements == 0 && !hasReshapes) {
           // Use size-1 vectors to increase parallelism if larger ones causes
           // idle threads in the subgroup.
           bool hasIdleThreads = distInfo.partitionableLoops.size() == 1 &&
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
index 9ccaa237cf69..51c192e371cd 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
@@ -11,6 +11,7 @@
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
 #include "mlir/IR/Operation.h"
 #include "mlir/Interfaces/FunctionInterfaces.h"
+#include "iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.h"
 
 namespace mlir::iree_compiler::IREE::GPU {
 
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
index 1d2cee51b8b2..3b38fe3e25b3 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/test/config_sort.mlir
@@ -83,7 +83,7 @@ func.func @sort5D_static_shape() {
 }
 
 //       CHECK: #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<pipeline = LLVMGPUTileAndFuse workgroup_size = [64, 1, 1] subgroup_size = 32>
-//       CHECK-DAG: func.func @sort5D_static_shape()
+//   CHECK-DAG: func.func @sort5D_static_shape()
 //       CHECK:     translation_info = #[[$TRANSLATION]]
 //       CHECK:   iree_linalg_ext.sort {
 //  CHECK-SAME:       lowering_config = #iree_gpu.lowering_config<{thread = [0, 1, 1, 1, 1], workgroup = [0, 1, 1, 1, 1]}>

From 312d27e17b511be513bf08e460639a2083903480 Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Mon, 7 Apr 2025 15:39:27 -0500
Subject: [PATCH 09/10] updating .bazel and CMakeLists.txt with dependency

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 .../iree/compiler/Codegen/Dialect/GPU/TargetUtils/BUILD.bazel    | 1 +
 .../iree/compiler/Codegen/Dialect/GPU/TargetUtils/CMakeLists.txt | 1 +
 .../compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp     | 1 +
 .../iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h  | 1 -
 4 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/BUILD.bazel b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/BUILD.bazel
index 0da075d519e4..eeef8ce53964 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/BUILD.bazel
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/BUILD.bazel
@@ -25,6 +25,7 @@ iree_compiler_cc_library(
         "//compiler/src/iree/compiler/Codegen/Common/GPU:GPUHeuristics",
         "//compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR:IREECodegenDialect",
         "//compiler/src/iree/compiler/Codegen/Dialect/GPU/IR:IREEGPUDialect",
+        "//compiler/src/iree/compiler/Codegen/Interfaces:PartitionableLoopsInterface",
         "//compiler/src/iree/compiler/Codegen/Utils",
         "//compiler/src/iree/compiler/Dialect/LinalgExt/Utils",
         "@llvm-project//llvm:Support",
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/CMakeLists.txt b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/CMakeLists.txt
index e873f641f202..29e9cd2b3579 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/CMakeLists.txt
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/CMakeLists.txt
@@ -28,6 +28,7 @@ iree_cc_library(
     iree::compiler::Codegen::Common::GPU::GPUHeuristics
     iree::compiler::Codegen::Dialect::Codegen::IR::IREECodegenDialect
     iree::compiler::Codegen::Dialect::GPU::IR::IREEGPUDialect
+    iree::compiler::Codegen::Interfaces::PartitionableLoopsInterface
     iree::compiler::Codegen::Utils
     iree::compiler::Dialect::LinalgExt::Utils
   PUBLIC
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
index 5c24338866fc..2999c32991f1 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp
@@ -14,6 +14,7 @@
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUEnums.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h"
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.h"
 #include "iree/compiler/Codegen/Utils/Utils.h"
 #include "iree/compiler/Dialect/LinalgExt/Utils/Utils.h"
 #include "llvm/ADT/STLExtras.h"
diff --git a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
index 51c192e371cd..9ccaa237cf69 100644
--- a/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
+++ b/compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h
@@ -11,7 +11,6 @@
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
 #include "mlir/IR/Operation.h"
 #include "mlir/Interfaces/FunctionInterfaces.h"
-#include "iree/compiler/Codegen/Interfaces/PartitionableLoopsInterface.h"
 
 namespace mlir::iree_compiler::IREE::GPU {
 

From 4cedafb4d8c95313a8530835cfca49cb8ef762df Mon Sep 17 00:00:00 2001
From: Muzammiluddin Syed <muzasyed@amd.com>
Date: Wed, 9 Apr 2025 16:31:26 -0500
Subject: [PATCH 10/10] Make workgroupSpace the default address space for alloc

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>
---
 compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp b/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
index 5cc0b8e9fb63..67f9ce5f97c8 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp
@@ -306,6 +306,8 @@ static FailureOr<Value> gpuRequireMemSpaceAllocationFn(OpBuilder &builder,
   MemRefType allocType = memRefType;
   auto privateSpace = gpu::AddressSpaceAttr::get(
       builder.getContext(), gpu::GPUDialect::getPrivateAddressSpace());
+  auto workgroupSpace = gpu::AddressSpaceAttr::get(
+      builder.getContext(), gpu::GPUDialect::getWorkgroupAddressSpace());
   if (!memorySpace) {
     allocType =
         MemRefType::get(memRefType.getShape(), memRefType.getElementType(),
@@ -317,6 +319,9 @@ static FailureOr<Value> gpuRequireMemSpaceAllocationFn(OpBuilder &builder,
     return builder.create<memref::AllocaOp>(loc, allocType, dynamicSizes)
         .getResult();
   }
+  allocType =
+      MemRefType::get(memRefType.getShape(), memRefType.getElementType(),
+                      AffineMap(), workgroupSpace);
   return builder.create<memref::AllocOp>(loc, allocType, dynamicSizes)
       .getResult();
 }
